### PR TITLE
Decouple subscription setup with StreamProvider set up

### PR DESCRIPTION
**Pull request by xiazen**
Programmatic Subscribe round2:

Decouple subscription setup with StreamProvider set up
Add tests to test it

### CODE REVIEW

File: src/Orleans/Streams/Core/IStreamSubscriptionManagerAdmin.cs [hunk 1]
```diff
@@ -8,6 +8,11 @@ namespace Orleans.Streams.Core
 {
     public interface IStreamSubscriptionManagerAdmin
     {
-        IStreamSubscriptionManager GetStreamSubscriptionManager(string name);
+        IStreamSubscriptionManager GetStreamSubscriptionManager(StreamSubscriptionManagerType managerType);
```
Comment (jason-bragg) [inline src/Orleans/Streams/Core/IStreamSubscriptionManagerAdmin.cs]: I think the type should remain a string, as having it as a defined enum means we can't add new ones without code change.  Limits extensibility.
We can have defined string values for the natively supported types.

File: src/Orleans/Core/ClientBuilder.cs [hunk 2]
```diff
@@ -86,6 +87,8 @@ private static void AddBasicServices(IServiceCollection services, ClientConfigur
             services.AddSingleton<StreamProviderManager>();
             services.AddSingleton<ClientStatisticsManager>();
             services.AddFromExisting<IStreamProviderManager, StreamProviderManager>();
+            services.AddFromExisting<IStreamProviderRuntime, ClientProviderRuntime>();
+            services.AddSingleton<IStreamSubscriptionManagerAdmin>(sp => new StreamSubscriptionManagerAdmin(sp.GetRequiredService<IStreamProviderRuntime>()));
```
Comment (jason-bragg) [inline src/Orleans/Core/ClientBuilder.cs]: I don't think you need to use the service provider to look up the runtime, constructor injection should do this for you with:
services.AddSingleton<IStreamSubscriptionManagerAdmin,StreamSubscriptionManagerAdmin>();

File: src/Orleans/Streams/Core/StreamSubscriptionManagerAdmin.cs [hunk 3]
```diff
@@ -8,22 +8,27 @@ namespace Orleans.Streams.Core
...
     {
-        private IStreamProviderManager providerManager;
-        public StreamSubscriptionManagerAdmin(IStreamProviderManager providerManager)
+        private IStreamSubscriptionManager explicitStreamSubscriptionManager;
+        public StreamSubscriptionManagerAdmin(IStreamProviderRuntime providerRuntime)
         {
-            this.providerManager = providerManager;
+            // using ExplicitGrainBasedAndImplicit pubsub here, so if StreamSubscriptionManager.Add(Remove)Subscription called on a implicit subscribed
+            // consumer grain, its subscription will be handled by ImplicitPubsub, and will not be messed into GrainBasedPubsub 
+            this.explicitStreamSubscriptionManager = new StreamSubscriptionManager(providerRuntime.PubSub(StreamPubSubType.ExplicitGrainBasedAndImplicit),
```
Comment (jason-bragg) [inline src/Orleans/Streams/Core/StreamSubscriptionManagerAdmin.cs]: Suggest adding StreamSubscriptionManager to a dictionary by name, and looking up manager by name in GetStreamSubscriptionManager.

File: src/Orleans/Streams/Core/StreamSubscriptionManagerAdmin.cs [hunk 4]
```diff
@@ -8,22 +8,27 @@ namespace Orleans.Streams.Core
...
-            var provider = this.providerManager.GetStreamProvider(providerName);
-            if (provider is IStreamSubscriptionManagerRetriever)
+            IStreamSubscriptionManager manager = null;
+            switch (managerType)
             {
-                return ((IStreamSubscriptionManagerRetriever)provider).GetStreamSubscriptionManager();
+                case StreamSubscriptionManagerType.ExplicitSubscribeOnly:
+                    manager = this.explicitStreamSubscriptionManager;
+                    break;
+                default: manager = this.explicitStreamSubscriptionManager;
```
Comment (jason-bragg) [inline src/Orleans/Streams/Core/StreamSubscriptionManagerAdmin.cs]: If passed in type is not found, we should throw, right?

**Comment by xiazen** [comment]
all good suggestions. addressed

**Comment by jdom** [comment]
Tests are randomly failing :( https://ci.dot.net/job/dotnet_orleans/job/master/job/netfx_prtest/870/

**Pull request merged**